docs/Toy/toy/dialect.py

```diff
@@ -58,22 +71,24 @@ class AddOp(Operation):
     The shapes of the tensor operands are expected to match.
     """
     name: str = 'toy.add'
-    arguments: Annotated[VarOperand, AnyTensorTypeF64]
+    lhs: Annotated[Operand, AnyTensorTypeF64]
+    rhs: Annotated[Operand, AnyTensorTypeF64]
     res: Annotated[OpResult, AnyTensorTypeF64]
 
     @classmethod
     def from_summands(cls: type[AddOp], lhs: SSAValue, rhs: SSAValue) -> AddOp:
-        assert isinstance(lhs.typ, TensorType | UnrankedTensorType)
-        element_type = cast(Float64Type,
-                            cast(TensorType[Any], lhs.typ).element_type)
-        return cls.create(result_types=[element_type], operands=[lhs, rhs])
+        assert isa(lhs.typ, AnyTensorTypeF64)
+        if isinstance(lhs.typ, TensorType):
+            result_typ = lhs.typ
+        else:
+            result_typ = rhs.typ
+        return cls.create(result_types=[result_typ], operands=[lhs, rhs])
```

Comment:
I don't like using `cls` very much. Do you needs this constructor in subclasses or is it unnecesssary?

Reply:
I kind of like it, classmethods should use `cls`, even if they don't have subclasses, it's the most flexible option. My hope is that we can transition to inits over the summer, and then we wouldn't have to worry about it anyway, but that's the API that I like the most for the time being.

Reply:
Thought about it some more, changing